src/libraries/System.Private.CoreLib/src/System/Collections/Generic/NonRandomizedStringEqualityComparer.cs

```diff
@@ -12,23 +13,71 @@ namespace System.Collections.Generic
     // randomized string hashing.
     [Serializable] // Required for compatibility with .NET Core 2.0 as we exposed the NonRandomizedStringEqualityComparer inside the serialization blob
     // Needs to be public to support binary serialization compatibility
-    public sealed class NonRandomizedStringEqualityComparer : EqualityComparer<string?>, ISerializable
+    public sealed class NonRandomizedStringEqualityComparer : INonRandomizedEqualityComparer<string?>, ISerializable
     {
-        internal static new IEqualityComparer<string?> Default { get; } = new NonRandomizedStringEqualityComparer();
+        internal static readonly IEqualityComparer<string?> AroundDefaultComparer = new NonRandomizedStringEqualityComparer(wrapsOrdinalComparer: false);
+        internal static readonly IEqualityComparer<string?> AroundStringComparerOrdinal = new NonRandomizedStringEqualityComparer(wrapsOrdinalComparer: true);
 
-        private NonRandomizedStringEqualityComparer() { }
+        // Flag indicates whether this instance wraps EqualityComparer<string>.Default
+        // or StringComparer.Ordinal.
+        private readonly bool _wrapsOrdinalComparer;
```

Comment:
I was going to guess this makes the dictionaries one pointer bigger. But I learned something new to me, it seems in at least some cases the CLR will pack boolean fields into the object header, apparently:
```c#
using System;
class Program
{
    static void Main(string[] args)
    {
        int size = 50000;
        X[] array = new X[size];
        long before = GC.GetTotalMemory(true);
        for (int i = 0; i < array.Length; i++)
        {
            array[i] = new X();
        }
        long after = GC.GetTotalMemory(true);

        Console.WriteLine((after - before)/ size);

        GC.KeepAlive(array);
    }
}

class X // comment out all bools, and it prints 24 (on 64 bit)
{
    bool a;
    bool b;
    bool c;
    bool d;
    bool e;
    bool f;
    bool g;
    bool h; // with bools a-h it still shows 24
    // bool i; // uncomment this and it shows 32
}
```

Reply:
@jkotas do I understand that right or am I misreading what I did?

Reply:
The minimum object size with CoreCLR GC is 12 bytes on 32-bit platforms and 24 bytes on 64-bit platforms. If the payload is less, the object size is padded to the minimum size.

Reply:
Ah - padding, of course. Thank you.

Reply:
It's on the singleton comparer instance, not the dictionary instance. That's why there are two singletons for the non-randomized equality comparer: one that wraps `EqualityComparer<string>.Default`, and one that wraps `StringComparer.Ordinal`. Their runtime behavior is identical. The only difference is that for compat I need to make sure that when you call the `Dictionary<...>.Comparer` property getter I give you back the same _comparer_ instance you used to instantiate the dictionary. So this information is smuggled via the comparer.

Reply:
To answer your other question re: how bools are stored, all objects in the CLR are at least 3 pointers (24 bytes on x64) in size. The first two pointers are collectively referred to as the "object header". The object's member data always begins just after these two pointers.

Due to how the GC handles variable-length objects (`string`, `Array`, `Utf8String`), the _length_ field must always be the first thing directly after the object header. When walking objects, the GC always attempts to dereference this field, even for normal fixed-length objects which don't have a _length_ field. It'll eventually get ignored for these types of objects. But the interesting consequence to this is that _for all objects_ there must be a pointer's worth of dereferenceable data immediately after the object header. It can be junk, but it must be dereferenceable. For normal fixed-size objects, the VM starts putting the member data at where this _length_ field would have been.

That's why the object size doesn't change when you have between 0 - 8 bools, because they take up the space that was already carved out for this object's member data. But as soon as you have a 9th bool it spills over.

Reply:
@GrabYourPitchforks but that wouldn't require **each** object to to be padded to at least 12/24 bytes, only the current gc segment would need to be at least 4/8 byte longer to ensure that there's valid memory to dereference there. The primary reason i've read for why it is that way is (plug is a continous memory range of surviving objects and gap is a continous range of dead objects / free space)
![image](https://user-images.githubusercontent.com/7110884/81659146-bb9e9380-9439-11ea-9834-4565669fce8c.png)

Reply:
@Suchiman Sure, I suppose you could do this as an optimization, but why bother? The number of zero-length objects (basically, _System.Object_ instances and nothing else) allocated is vanishingly small compared to the number of "useful" object instances allocated. Doesn't seem worthwhile for the GC to change from its existing logic due since we'd expect no real-world gains.

Reply:
@GrabYourPitchforks this doesn't just affect zero-length objects but all objects that are smaller than 4/8 bytes, [e.g. a few bools](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWAChHqBmRgb0YAIevqruvTg16iuASTwBnAA4wwGAHQAJGAEMZACjwwA7lwDCmgJTGA3IJ4BfRjZFiHzKoY6W+LLsAgQANlwBCNBb2PBQeXr4BVMFWQA==), but the point i was trying to make, was that this is unlikely to be the cause but rather an effect (if the objects are always that large, one can save on the check) of the objects being padded. The likely primary cause is that the GC needs the additional space for the plan phase of the GC (as seen in the previous excerpt from the book).